### ISSUE TITLE
no-unused-vars needs a better default

### ISSUE THREAD

**Issue by xixixao** [eslint/eslint#14593]
Preface: `no-unused-vars` is one of the few rules that actually make me use ESLint in the world of Prettier and Snowpack etc. But if I use `eslint --init` I should get sensible defaults that let me write JS. Maybe not the defaults I prefer, but there should be a way to write all today's standard JS, and have workaround for each warning/error.

`no-unused-vars` forces me to dive into ESLint configuration because of the two patterns listed below.

**What rule do you want to change?**
no-unused-vars

**Does this change cause the rule to produce more or fewer warnings?**
Fewer.

**How will the change be implemented? (New option, new default behavior, etc.)?**
new default behavior

**Please provide some example code that this change will affect:**

```js
const {foo, ...bar} = bla;
const [_, bar] = bla;
```

**What does the rule currently do for this code?**

Errors on both

**What will the rule do after it's changed?**

Well, that depends. I think we have two options:

## Option A
Turn on `ignoreRestSiblings` to handle the first line, and suggest removing the variable for the second line (leading to `const [, bar] = bla`).

This is not my preferred style of JS, I think it's less readable / more confusing, but it works.

## Option B
Turn on `varsIgnorePattern: "^_"`, and suggest `const {foo: _, ...bar}` for the first line.

This is my preferred style of JS, which I think is widely adopted. I guess the only issue is clashing with underscore/lodash, on which I don't have much context. If that is the blocking issue, I would suggest:

## Option C
Turn on `varsIgnorePattern: "^_\w"`, where `\w` would be replaced by the correct JS allowed variable name character I cannot remember atm. This leads to even more readable code, but it can be annoying like on the second line example. `const {foo: _foo, ...bar} = bla;`, `const [_all, bar] = bla;`

I'd probably prefer Option B over C, but either is better than status quo.

Finally for either option B or C, I would suggest:

## Option D
(and this one is "breaking" and unlikely to make it) set `argsIgnorePattern` to the same thing as `varsIgnorePattern` and set `args` to `"all"`, because prefixed args are much more explicit than relying on later arguments being used.

While I prefer B and D, I can that this might not work for underscore users, so I would suggest letting users choose in `--init` between B&D and A.

**Are you willing to submit a pull request to implement this change?**
Sure.

**Comment by nzakas** [comment]
Thanks for the input. We can’t guarantee that the defaults in ESLint will work with all coding styles, so we choose the settings that we thing apply most universally. Obviously this won’t work for everyone but at least it’s a starting point.

I can see an argument that perhaps ignoreRestSiblings should be on be default, though I’m not sure it’s any more compelling than having it off by default.

For the second example, that’s more of a stylistic preference so we can’t support that by default. You’ll need to modify the rule configuration for that one.

**Comment by btmills** [comment]
I agree with @nzakas on this. I could possibly be convinced that `ignoreRestSiblings` could be on by default as in option A, but there are also strong arguments against it, so I give the benefit of the doubt to the status quo. Options B through D all prescribe a particular style of writing code. In cases where there's not a clear best practice, ESLint intentionally cedes stylistic decisions to end users and shared config authors. The `"^_"` pattern is common enough that the docs demonstrate it in the `argsIgnorePattern` example, but there's enough diversity in approaches that I don't see us being able to settle on a single de facto standard for a default.

**Comment by xixixao** [comment]
> I can see an argument that perhaps ignoreRestSiblings should be on be default, though I’m not sure it’s any more compelling than having it off by default.

Why?

> there are also strong arguments against it

Which ones? (curious)

> a particular style of writing code

Do you agree that the two lines in the issue are part of the common mechanisms of JS the language? (the ability to destructure and ignore either named fields or indexes in arrays)

If we agree on that, then I think there are two styles (for each line), where the `_` style obviously aligns well together. Right now ESLint somewhat favors one style (`const [, bar] = bla;`) with its defaults (and doesn't have a sane default for the field destructuring at all imho).

I get that the `--init` script shouldn't have 50 choices, but this (if we bundle it with function parameters) is a super common scenario that sooner or later every single codebase will have to deal with. (I think this is a good litmus test for what should be left to the configuration: "Is everyone using ESLint going to need this?" If yes, it should either have a good default or go into the `--init` script)

> We can’t guarantee that the defaults in ESLint will work with all coding styles, so we choose the settings that we thing apply most universally.

Is there any data on `_` for ignoring being more or less common than the other workarounds for the three cases (arrays, objects, function params)?

**Comment by xixixao** [comment]
Prettier does this well btw - you might not like its defaults, but there is no valid JavaScript you can write that will force you to go configure it.

**Comment by btmills** [comment]
Responding just to the portions replying to me:

> > there are also strong arguments against it
> 
> Which ones? (curious)

Enabling that option by default weakens the rule. It would be surprising if `no-unused-vars`, given `const { used, unused, alsoUnused, ...rest } = obj;`, misses `unused` and `alsoUnused` by default. Sure, those variables might be _less_ unused than other unused variables, but the rule can't tell without a hint whether they're actually unused or this was just a convenient syntax to remove those keys from `obj`. The options let developers provide that hint to match their preferred convention.

`no-unused-vars` is included in `eslint:recommended`, and those rules face a high bar for inclusion. False positives from rules in `eslint:recommended` should be rare. We only include rules whose reports are common mistakes, not opinion-based stylistic choices. If we were to change a default, we'd be making an opinionated decision to default to one convention.

> > a particular style of writing code
> 
> Do you agree that the two lines in the issue are part of the common mechanisms of JS the language?

Absolutely! This is a common use case in the language and one I encounter on a regular basis.

> If we agree on that, then I think there are two styles (for each line), where the `_` style obviously aligns well together.

I don't have data to back it up, but I agree the underscore style seems to be the most common. However, it's not always appropriate. For example, you astutely mentioned in the issue description that it could conflict with underscore/lodash. It would be hard to justify setting a default that could conflict with a package that gets ~2.5x as many downloads as ESLint itself.

> Prettier does this well btw - you might not like its defaults, but there is no valid JavaScript you can write that will force you to go configure it.

I agree, Prettier is a great time-saver! Formatting decisions can be made that way because it's rare for them to be objectively incorrect, just contrary to some portion of opinions. For `no-unused-vars`, any false positives are bugs, and by default, so are false negatives. The patterns that you've identified are common conveniences made even more so with destructuring syntax, so we have the options for developers who want to allow some false negatives for convenience. Even then, there are multiple ways to do that (`ignoreRestSiblings` being the most lax, `varsIgnorePattern` in the middle, and [disable comments](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOiAyICovXG5cbmV4cG9ydCBkZWZhdWx0IGNsYXNzIE15Q29tcG9uZW50IGV4dGVuZHMgUmVhY3QuQ29tcG9uZW50IHtcblxuXHQvLyAuLi5cblxuXHRyZW5kZXIoKSB7XG5cdFx0Y29uc3Qge1xuXHRcdFx0LyogZXNsaW50LWRpc2FibGUgbm8tdW51c2VkLXZhcnMgLS0gYGlkYCBpc24ndCB1c2VkIGhlcmUgaW5cblx0XHRcdCAgIGByZW5kZXIoKWAsIGJ1dCBpdCBpcyB1c2VkIGluIGhhbmRsZXJzLiBQdWxsIGl0IG91dCBzbyB0aGF0IHdlXG5cdFx0XHQgICBkb24ndCBzcHJlYWQgaXQgb250byB0aGUgYDxkaXY+YCB3aXRoIHRoZSByZXN0IG9mIHRoZSBwcm9wcy4gKi9cblx0XHRcdGlkLFxuXHRcdFx0LyogZXNsaW50LWVuYWJsZSBuby11bnVzZWQtdmFycyAqL1xuXHRcdFx0Li4ucHJvcHNcblx0XHR9ID0gdGhpcy5wcm9wcztcblxuXHRcdHJldHVybiA8ZGl2IHsuLi5wcm9wc30+ey8qIC4uLiAqL308L2Rpdj47XG5cdH1cbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjEyLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7ImpzeCI6dHJ1ZX19LCJydWxlcyI6e30sImVudiI6e319fQ==) the most strict), so how would we decide which convention to make the default?

**Comment by ljharb** [comment]
Using `_` to mean "ignore this" or "private" are both very common - that doesn't mean they're a good idea. It's *much* better to omit a binding when you can, than to create one that's only nonexistent by convention.

**Comment by nzakas** [comment]
Given that there’s no strong support for making a change here, I’m closing this issue. As I mentioned previously, someone will always believe our defaults are incorrect based on how they write their code, and while there is plenty of room for debate on what the most correct take is, that’s not a debate that’s useful for us to have for each rule.
